### ISSUE TITLE
S3 Uploader issue

### ISSUE THREAD

**Issue by petrovicboban** [HubSpot/Singularity#1432]
Hi,  
I have two tasks runnng on host A, the first task is constrained to host A.  
I remove all files from s3/metadata folder. When I bounce first task (restart). s3/metadata contains two files, one for the killed task and one for the running task. S3Uploader do what it has to do , and uploads rotated file of the killed task.  If I bounce the second task now, that creates new file in s3/metadata for it, but s3 uploader somehow is confused and thinks that new rotated log file in logs folder belongs to the first, killed, task, and uploads it that way, under the name of that first, killed task. Either something is wrong with s3uploader here, or with executor itself, because I dont know why it keeps both files for killed and running task in s3/metadata. After some time file for the old , killed task expire, but why that dont happen immidiatly?

**Comment by ssalinas** [comment]
A few things on the S3Uploader

In terms of the old metadata file not disappearing, the poller that checks these files runs every 10 mins by default (you can change this interval in the configuration). It will remove the file in one of two cases:
  - The metadata was marked as `finished` (done by the executor when the task finishes)
  - there have been no new files found to upload in the specified folder in a period of time

As for the duplicates, I want to make sure I understand the situation here. Are these separate tasks with separate executors in separate sandboxes? It would be helpful to see the S3Uploader logs, executor logs, and some of the metadata to help debug this.

**Comment by petrovicboban** [comment]
I will produce issue on clean test environment for weekend.  
For now, I find another issue:  
TaskA is long running  task, and when I run logrotate manually over logrotate config file for that task, created by executor, it works as expected: it truncates service.log in TaskA sandbox, and creates service.log-XXXXX.log.gz in logs directory. It seems that S3Uploader dont see that file at the moment. 
Task with ID oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default is scheduled task, and it runs every hour, it finishes after 2 minutes. Upon running, here is the part of S3Uploader log:  
```
INFO  [2017-02-23 17:02:04,404] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Reading /var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json
INFO  [2017-02-23 17:02:04,406] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Created new uploader SingularityS3Uploader [uploadMetadata=S3UploadMetadata [directory=/var/log/singularity/logs, fileGlob=service.log*.[gb]z*, s3Bucket=stratagem-singularity, s3KeyFormat=oncourt_stratagem/%Y/%m/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_%index-%s%fileext, finished=false, onFinishGlob=Optional.absent(), pid=Optional.absent(), s3AccessKey=**empty**, s3Secret=**empty**, finishedAfterMillisWithoutNewFile=Optional.absent()], metadataPath=/var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json]
INFO  [2017-02-23 17:02:04,406] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Reading /var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json
INFO  [2017-02-23 17:03:03,986] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Waiting on 1 future(s)
INFO  [2017-02-23 17:03:03,986] [tyS3Uploader-27] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploading 1 item(s)
INFO  [2017-02-23 17:03:04,152] [tyS3Uploader-27] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploading /var/log/singularity/logs/service.log-201702231487867620.log.gz to stratagem-singularity/oncourt_stratagem/2017/02/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_0-1487867600000.gz (size 5030775)
INFO  [2017-02-23 17:03:04,730] [tyS3Uploader-27] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploaded oncourt_stratagem/2017/02/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_0-1487867600000.gz in 00:00.579
INFO  [2017-02-23 17:03:04,731] [tyS3Uploader-27] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploaded 1 out of 1 item(s) in 00:00.745
INFO  [2017-02-23 17:03:04,731] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Found 1 items from 1 uploader(s) in 00:00.745
INFO  [2017-02-23 17:03:12,009] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Reading /var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json
INFO  [2017-02-23 17:03:12,010] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Toggling uploader SingularityS3Uploader [uploadMetadata=S3UploadMetadata [directory=/var/log/singularity/logs, fileGlob=service.log*.[gb]z*, s3Bucket=stratagem-singularity, s3KeyFormat=oncourt_stratagem/%Y/%m/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_%index-%s%fileext, finished=false, onFinishGlob=Optional.absent(), pid=Optional.absent(), s3AccessKey=**empty**, s3Secret=**empty**, finishedAfterMillisWithoutNewFile=Optional.absent()], metadataPath=/var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] finish state to true
INFO  [2017-02-23 17:03:12,010] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Reading /var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json
INFO  [2017-02-23 17:03:12,010] [main] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Toggling uploader SingularityS3Uploader [uploadMetadata=S3UploadMetadata [directory=/var/log/singularity/logs, fileGlob=service.log*.[gb]z*, s3Bucket=stratagem-singularity, s3KeyFormat=oncourt_stratagem/%Y/%m/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_%index-%s%fileext, finished=false, onFinishGlob=Optional.absent(), pid=Optional.absent(), s3AccessKey=**empty**, s3Secret=**empty**, finishedAfterMillisWithoutNewFile=Optional.absent()], metadataPath=/var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] finish state to true
INFO  [2017-02-23 17:13:03,986] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Waiting on 1 future(s)
INFO  [2017-02-23 17:13:03,986] [tyS3Uploader-28] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploading 1 item(s)
INFO  [2017-02-23 17:13:04,139] [tyS3Uploader-28] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploading /var/log/singularity/logs/service.log-201702231487869391.log.gz to stratagem-singularity/oncourt_stratagem/2017/02/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_0-1487869391000.gz (size 3974)
INFO  [2017-02-23 17:13:04,440] [tyS3Uploader-28] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploaded oncourt_stratagem/2017/02/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_0-1487869391000.gz in 00:00.301
INFO  [2017-02-23 17:13:04,441] [tyS3Uploader-28] c.h.singularity.s3uploader.SingularityS3Uploader 18018 - [oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json] Uploaded 1 out of 1 item(s) in 00:00.455
INFO  [2017-02-23 17:13:04,441] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Found 1 items from 1 uploader(s) in 00:00.455
INFO  [2017-02-23 17:23:03,986] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Waiting on 1 future(s)
INFO  [2017-02-23 17:23:03,986] [arityS3Driver-0] c.h.s.s3uploader.SingularityS3UploaderDriver 18018 - Expiring SingularityS3Uploader [uploadMetadata=S3UploadMetadata [directory=/var/log/singularity/logs, fileGlob=service.log*.[gb]z*, s3Bucket=stratagem-singularity, s3KeyFormat=oncourt_stratagem/%Y/%m/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT_%index-%s%fileext, finished=false, onFinishGlob=Optional.absent(), pid=Optional.absent(), s3AccessKey=**empty**, s3Secret=**empty**, finishedAfterMillisWithoutNewFile=Optional.absent()], metadataPath=/var/log/singularity/s3/metadata/oncourt_stratagem-oncourt_stratagem_deploy_1-1487869322306-1-db05.stratagem.co-DEFAULT-default.s3.json]

```

As it can be seen, S3uploader sees both logrotated file of TaskA and logrotated file of scheduled task, as log files of scheduled task and as such, it uploads them to S3. So, scheduled, just finished task ends with two log files, one mistakenly uploaded which should belong to TaskA.

**Comment by ssalinas** [comment]
Interesting part there is `directory=/var/log/singularity/logs`. This won't work if you are rotating all of your logs to a single folder. It was coded with the intent that each would be  in a folder in the task sandbox. The matcher is just a glob for files in a folder. So, if multiple service.log files are in the same folder, they will be uploaded.

Are all of your tasks rotating logs to some central folder?

**Comment by petrovicboban** [comment]
Hmm, yes. That makes sense.  
I have this in executor config:  
```
logrotateToDirectory: /var/log/singularity/logs
````

So if I set it as relative path, it will be relative to what? Sandbox folder?

**Comment by ssalinas** [comment]
Correct, which is why the default is `logrotateToDirectory: logs`. That way not only are the logs separated, but they are easily accessible in the sandbox for the individual task (which is then visible in the SingularityUI as well)

**Comment by petrovicboban** [comment]
Thanks a lot for explanation!
